File: solver/src/main/java/will/sudoku/solver/StepRecorder.kt
package will.sudoku.solver

import will.sudoku.solver.Board
import will.sudoku.solver.Coord

/**
 * Records each step of the solving process.
 *
 * Tracks:
 * - Each cell filled
 * - Each backtracking event
 * - Each eliminator application
 * - Complete solving progress
 *
 * ## Usage
 *
 * ```kotlin
 * val solver = Solver()
 * val recorder = StepRecorder()
 * val solution = solver.solve(board, recorder)
 *
 * if (solution != null) {
 *     println("Solved in ${recorder.progress.stepCount} steps")
 *     recorder.progress.steps.forEach { step ->
 *         println("${step.stepNumber}: ${step.explanation}")
 *     }
 * }
 * ```
 *
 * @see SolvingProgress for the complete progress data
 * @see SolvingStep for individual step details
 */
class StepRecorder : SolvingListener {
    
    private val _steps = mutableListOf<SolvingStep>()
    private var stepNumber = 0
    private var currentBoardState: String? = null
    private var solved = false
    private var noSolution = false
    private var noSolutionReason: String? = null
    
    /**
     * The solving progress recorded so far.
     */
    val progress: SolvingProgress
        get() {
            val progressObj = SolvingProgress(
                originalPuzzle = currentBoardState ?: "",
                currentBoardState = currentBoardState ?: ""
            )
            _steps.forEach { progressObj.addStep(it) }
            if (solved) {
                progressObj.markSolved(currentBoardState ?: "")
            } else if (noSolution) {
                progressObj.markNoSolution(noSolutionReason ?: "No valid solution found")
            }
            return progressObj
        }
    
    override fun onCellFilled(coord: Coord, value: Int, explanation: String) {
        stepNumber++
        _steps.add(SolvingStep.cellFilled(
            stepNumber = stepNumber,
            cell = coord,
            value = value,
            explanation = explanation,
            boardState = currentBoardState ?: ""
        ))
    }
    
    override fun onBacktracking() {
        stepNumber++
        _steps.add(SolvingStep.backtrack(
            stepNumber = stepNumber,
            cell = Coord(0, 0), // Placeholder - no specific cell for backtracking
            wrongValue = 0,     // Placeholder - no specific value
            explanation = "Backtracking - no valid candidates found"
        ))
    }
    
    override fun onEliminatorApplied(name: String, eliminations: Int) {
        // Per-cell detail is handled by onCandidatesEliminated.
        // This aggregate callback is kept for backward compatibility with other listeners.
    }

    override fun onCandidatesEliminated(name: String, eliminations: List<Elimination>) {
        if (eliminations.isEmpty()) return

        stepNumber++
        val affectedCells = eliminations.map { it.cell }
        val allValues = eliminations.flatMap { it.eliminatedValues }.toSet()
        val totalEliminated = eliminations.sumOf { it.eliminatedValues.size }

        // Build human-readable per-cell explanation
        val cellDescs = eliminations.sortedBy { "${it.cell.row},${it.cell.col}" }
            .joinToString("; ") { e ->
                val cellLabel = "(${e.cell.row + 1}, ${e.cell.col + 1})"
                val vals = e.eliminatedValues.sorted().joinToString(",")
                "$cellLabel: [$vals]"
            }

        val stepType = StepType.fromTechniqueName(name)
        val explanation = "$name: $totalEliminated candidate(s) eliminated — $cellDescs"

        _steps.add(SolvingStep(
            stepNumber = stepNumber,
            stepType = stepType,
            affectedCells = affectedCells,
            values = allValues,
            explanation = explanation
        ))
    }
    
    override fun onSolveComplete(solved: Boolean, timeNanos: Long, backtracks: Int) {
        if (solved) {
            this.solved = true
        } else {
            this.noSolution = true
            this.noSolutionReason = "No valid solution found"
        }
    }
    
    override fun onPropagationPassStarted() {
        // No action needed
    }
    
    override fun onPropagationPassComplete(cellsFilled: Int, candidatesEliminated: Int) {
        // No action needed
    }
    
    override fun onGuessMade(coord: Coord, value: Int, candidateCount: Int) {
        // Already handled by onCellFilled with "Guess" explanation
    }
    
    /**
     * Sets the current board state (called by solver).
     */
    fun setBoardState(boardString: String) {
        currentBoardState = boardString
    }
}

File: packages/solver/src/StepRecorder.ts
import type { Board } from './Board'
import type { Coord } from './Coord'
import type { SolvingListener, Elimination } from './Solver'
import { StepType, stepTypeFromTechniqueName } from './StepType'
import type { SolvingStep } from './SolvingStep'
import { cellFilled, backtrack as backtrackStep } from './SolvingStep'
import { SolvingProgress } from './SolvingProgress'

/**
 * Records each step of the solving process.
 *
 * Implements SolvingListener to capture every cell fill, backtracking event,
 * and eliminator application into a SolvingProgress.
 *
 * Equivalent to the Kotlin StepRecorder class.
 *
 * @example
 * ```ts
 * const solver = new Solver()
 * const recorder = new StepRecorder()
 * const solution = solver.solve(board, recorder)
 * if (solution) {
 *   console.log(`Solved in ${recorder.progress.steps.length} steps`)
 * }
 * ```
 */
export class StepRecorder implements SolvingListener {
  private _steps: SolvingStep[] = []
  private _stepNumber = 0
  private _currentBoardState = ''
  private _solved = false
  private _noSolution = false
  private _noSolutionReason = ''

  /** The solving progress recorded so far. */
  get progress(): SolvingProgress {
    const progressObj = new SolvingProgress(this._currentBoardState)
    for (const step of this._steps) {
      progressObj.addStep(step)
    }
    if (this._solved) {
      progressObj.markSolved(this._currentBoardState)
    } else if (this._noSolution) {
      progressObj.markNoSolution(this._noSolutionReason)
    }
    return progressObj
  }

  /** Set the current board state (called by solver). */
  setBoardState(boardString: string): void {
    this._currentBoardState = boardString
  }

  onCellFilled(coord: Coord, value: number, explanation: string): void {
    this._stepNumber++
    this._steps.push(cellFilled(
      this._stepNumber,
      coord,
      value,
      explanation,
      this._currentBoardState
    ))
  }

  onBacktracking(): void {
    this._stepNumber++
    this._steps.push(backtrackStep(
      this._stepNumber,
      { row: 0, col: 0, index: 0 }, // Placeholder — no specific cell for backtracking
      0, // Placeholder — no specific value
      'Backtracking — no valid candidates found'
    ))
  }

  onCandidatesEliminated(techniqueName: string, eliminations: Elimination[]): void {
    if (eliminations.length === 0) return

    this._stepNumber++
    const affectedCells = eliminations.map(e => e.coord)
    const allValues = new Set<number>()
    let totalEliminated = 0
    for (const e of eliminations) {
      for (const v of e.eliminatedValues) {
        allValues.add(v)
        totalEliminated++
      }
    }

    // Build human-readable per-cell explanation
    const cellDescs = [...eliminations]
      .sort((a, b) => a.coord.row * 9 + a.coord.col - (b.coord.row * 9 + b.coord.col))
      .map(e => {
        const cellLabel = `(${e.coord.row + 1},${e.coord.col + 1})`
        const vals = e.eliminatedValues.sort().join(',')
        return `${cellLabel}: [${vals}]`
      })
      .join('; ')

    const stepType = stepTypeFromTechniqueName(techniqueName)
    const explanation = `${techniqueName}: ${totalEliminated} candidate(s) eliminated — ${cellDescs}`

    this._steps.push({
      stepNumber: this._stepNumber,
      stepType,
      affectedCells,
      values: allValues,
      explanation,
      timestamp: Date.now()
    })
  }

  onSolveComplete(solved: boolean, _timeNanos: number, _backtracks: number): void {
    if (solved) {
      this._solved = true
    } else {
      this._noSolution = true
      this._noSolutionReason = 'No valid solution found'
    }
  }

  // No-op callbacks for optional listener methods
  onPropagationPassStarted(): void {}
  onEliminatorApplied(_techniqueName: string, _totalEliminated: number): void {}
  onGuessMade(_coord: Coord, _firstCandidate: number, _totalCandidates: number): void {}
}
